listener/listener.go

```diff
@@ -300,80 +315,150 @@ func (l *Listener) Stream(ctx context.Context) error {
 		},
 	}
 
-	tx := NewWalTransaction(l.log, pool, l.monitor)
-
-	for {
-		if err := ctx.Err(); err != nil {
-			l.log.Warn("stream: context canceled", "err", err)
-			return nil
-		}
-
-		msg, err := l.replicator.WaitForReplicationMessage(ctx)
-		if err != nil {
-			return fmt.Errorf("wait for replication message: %w", err)
-		}
-
-		if msg == nil {
-			l.log.Debug("got empty message")
-			continue
-		}
-
-		if err = l.processMessage(ctx, msg, tx); err != nil {
-			return fmt.Errorf("process message: %w", err)
+	topicSet := make(map[string]bool)
+	defer func() {
+		for topic := range topicSet {
+			l.publisher.Flush(topic)
 		}
+	}()
 
-		l.processHeartBeat(msg)
-	}
-}
+	tx := NewWalTransaction(l.log, pool, l.monitor)
 
-func (l *Listener) processMessage(ctx context.Context, msg *pgx.ReplicationMessage, tx *WalTransaction) error {
-	if msg.WalMessage == nil {
-		l.log.Debug("empty wal-message")
-		return nil
-	}
+	group, ctx := errgroup.WithContext(ctx)
+	messageChan := make(chan *pgx.ReplicationMessage, 100)
+	eventsChan := make(chan *messageAndEvents, 100)
+	resultChan := make(chan *eventAndPublishResult, 500)
```

Comment:
These numbers are super arbitrary.

The reason I made the last buffer 5x the other ones is that one message can explode into several events.